yt/data_objects/selection_data_containers.py

```diff
@@ -930,20 +930,78 @@ def _cond_ind(self):
+                         parent[(ptype, "particle_position_y")],
+                         parent[(ptype, "particle_position_z")]], axis=1).value
+        levels = self[("index", "grid_level")].astype('int32').value
+        levelmin = levels.min()
+        levelmax = levels.max()
+
+        mask = np.zeros(ppos.shape[0], dtype=bool)
+
+        for lvl in range(levelmax, levelmin-1, -1):
+            # Filter out cells not in the current level
+            lvl_mask = (levels == lvl)
+            dx_loc = dx[lvl_mask]
+            pos_loc = pos[lvl_mask]
+
+            grid_tree = _scipy.spatial.cKDTree(pos_loc, boxsize=1)
+
+            # Compute closest cell for all remaining particles
+            dist, icell = grid_tree.query(ppos[~mask], distance_upper_bound=dx_loc.max(),
+                                          p=np.inf)
+            mask_loc = np.isfinite(dist[:])
+
+            # Check that particles within dx of a cell are in it
+            i = icell[mask_loc]
+            dist = np.abs(ppos[~mask][mask_loc, :] - pos_loc[i])
+            tmp_mask = np.all(dist <= (dx_loc[i] / 2), axis=1)
+
+            mask_loc[mask_loc] = tmp_mask
+
+            # Update the particle mask with particles found at this level
+            mask[~mask] |= mask_loc
+
+        return mask
+
+    def _part_ind_brute_force(self, ptype):
+        parent = getattr(self, "parent", self.base_object)
+        units = "code_length"
+        mask = points_in_cells(
+            self[("index", "x")].to(units),
+            self[("index", "y")].to(units),
+            self[("index", "z")].to(units),
+            self[("index", "dx")].to(units),
+            self[("index", "dy")].to(units),
+            self[("index", "dz")].to(units),
+            parent[(ptype, "particle_position_x")].to(units),
+            parent[(ptype, "particle_position_y")].to(units),
+            parent[(ptype, "particle_position_z")].to(units))
+
+        return mask
+
     def _part_ind(self, ptype):
         if self._particle_mask.get(ptype) is None:
-            parent = getattr(self, "parent", self.base_object)
-            units = "code_length"
-            mask = points_in_cells(
-                self[("index", "x")].to(units),
-                self[("index", "y")].to(units),
-                self[("index", "z")].to(units),
-                self[("index", "dx")].to(units),
-                self[("index", "dy")].to(units),
-                self[("index", "dz")].to(units),
-                parent[(ptype, "particle_position_x")].to(units),
-                parent[(ptype, "particle_position_y")].to(units),
-                parent[(ptype, "particle_position_z")].to(units))
+            # If scipy is installed, use the fast KD tree
+            # implementation. Else, fall back onto the direct
+            # brute-force algorithm.
+            try:
+                _scipy.spatial.KDTree
+                mask = self._part_ind_KDTree(ptype)
```

Comment:
It might be worth emitting a logging message suggesting to install scipy. Please feel free to submit an issue or subsequent pull request with that if you feel like it, not a big deal.